nixos/modules/services/monitoring/datadog-agent.nix

```diff
@@ -0,0 +1,205 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.datadog-agent;
+
+  ddConf = {
+    dd_url              = "https://app.datadoghq.com";
+    skip_ssl_validation = "no";
+    api_key             = "";
+    collector_log_file  = "/var/log/datadog/collector.log";
+    forwarder_log_file  = "/var/log/datadog/forwarder.log";
+    dogstatsd_log_file  = "/var/log/datadog/dogstatsd.log";
+    pup_log_file        = "/var/log/datadog/pup.log";
+    confd_path          = "/etc/datadog-agent/conf.d";
+    additional_checksd  = "/etc/datadog-agent/checks.d";
+    use_dogstatsd       = "yes";
+  }
+  // optionalAttrs (cfg.logLevel != null) { log_level = cfg.logLevel; }
+  // optionalAttrs (cfg.hostname != null) { inherit (cfg) hostname; }
+  // optionalAttrs (cfg.tags != null ) { tags = concatStringsSep ", " cfg.tags; }
+  // cfg.extraConfig;
+
+  makeConfigDir = entries: mapAttrsToList (name: conf: {
+    source = pkgs.writeText (baseNameOf name) (builtins.toJSON conf);
+    target = "datadog-agent/" + name;
+  }) (filterAttrs (name: conf: conf != null) entries);
+
+  etcfiles = makeConfigDir
```

Comment:
The module in its current form lets users specify the configuration in the structure expected by Datadog directly (i.e. there's no "fancy" translations from a setting to monitor a list of URLs to `http` checks) - that's good because it doesn't introduce another extra layer that needs to be learned.

However, I think it'd be nicer if arbitrary check configurations could be added more easily instead of having a fixed set of supported check configurations here. Currently one needs to link them via `environment.etc` in addition and then extend the restart trigger of the agent's unit.

A suggestion would be to accept an attribute set of configurations & their structure, for example like this:

```nix
services.datadog-agent.configs = {
  http_check = {
    init_config = null;
    instances = [
      # ... and so on
    ];
  };

  disk = {
    # you get the gist of it
  };
};
```

The straightforward way to implement it would be to link those to `conf.d/${key}.yaml`, but I see that some checks have their own `.d` folder - is that required? If yes, the config attribute set could probably use the whole relative path from `conf.d/` as the base as its key.

Does this make sense? :)

Reply:
> is that required?

The answer is no. Datadog say that they prefer the `.d`-folder approach, but both work - so it's just a matter of choosing one.